### PR TITLE
:sparkles: Add leaderboard system

### LINKED ISSUE #215: [任務] 排行榜 - 查詢自己
### 任務描述
在 WSA Discord 頻道中，輸入 `/utopia leaderboard my-rank` 時，Server 會依照所有人的等級、經驗值、賞金及其他屬性，根據公式總和上述屬性，排列分數，並將輸入其指令的使用者名次，以文字呈現名次結果至 WSA 頻道中。

## Slash Command 資訊
1. command name：`utopia leaderboard`
2. command arguments：`my-rank`
3. command description：utopia 排行榜 - 查詢自己

### 測試步驟

輸入查詢自己排名指令

- Given
    - 資料庫有３名 player
        | Name | Lv | Exp | Bounty |
        | --- | --- | --- | --- |
        | self | 5 | 1870 | 155 |
        | Wally  | 4 | 1100 | 70 |
        | Mily  | 7 | 2912 | 14000 |

- When
    - 下指令 `/utopia leaderboard my-rank`

- Then
    - Output: `你的排名為第 2 名`

    - 示意圖
Screen Shot 2023-10-30 at 1 12 40 PM

### 注意事項
無

### PR DESCRIPTION
**Pull request by w305jack**
## Why need this change? / Root cause: 
- add leaderboard system feature
## Changes made:
- resolved #215 #216 
## Test Scope / Change impact:
- leaderboard system

### CODE REVIEW

File: utopia-gamification/src/main/kotlin/tw/waterballsa/utopia/utopiagamification/leaderboard/LeaderBoardListener.kt [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            )
+        } else {
+            buttons.add(
+                Button.success(
+                    makeButtonId(LEADERBOARD_NEXT_BUTTON, page.getPageable()),
+                    LEADERBOARD_NEXT_STRING
+                ).asDisabled()
+            )
+        }
+        return buttons
```
Comment (Wally5077) [inline utopia-gamification/src/main/kotlin/tw/waterballsa/utopia/utopiagamification/leaderboard/LeaderBoardListener.kt:158]:
```kotlin
val previousPageButton = createPreviousPageButton(page)
val nextPageButton = createNextPageButton(page)
return mutableListOf(previousPageButton, nextPageButton)
```

File: utopia-gamification/src/main/kotlin/tw/waterballsa/utopia/utopiagamification/leaderboard/LeaderBoardListener.kt [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    private fun Button.toPageable(): Pageable {
+        val splitButtonIds = id.toString().split("_")
+        return when (splitButtonIds.size) {
+            3 -> {
+                PageRequest.of(splitButtonIds[1].toInt(), splitButtonIds[2].toInt())
+            }
+            else -> {
+                PageRequest.ofSize(10)
+            }
+        }
```
Comment (Wally5077) [inline utopia-gamification/src/main/kotlin/tw/waterballsa/utopia/utopiagamification/leaderboard/LeaderBoardListener.kt:117]: destruct

File: utopia-gamification/src/main/kotlin/tw/waterballsa/utopia/utopiagamification/leaderboard/LeaderBoardListener.kt [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                    LEADERBOARD_NEXT_STRING
+                ).asDisabled()
+            )
+        }
+        return buttons
+    }
+
+    private fun makeButtonId(prefix: String, pageable: Pageable): String {
+        return "${prefix}_${pageable.getPageNumber()}_${pageable.getPageSize()}"
+    }
```
Comment (Wally5077) [inline utopia-gamification/src/main/kotlin/tw/waterballsa/utopia/utopiagamification/leaderboard/LeaderBoardListener.kt:163]: `createPageButtonId`

**Review by Wally5077 — CHANGES_REQUESTED** [review]
**Review done.**

**Review by Wally5077 — APPROVED** [review]
**LGTM**

**Review by m1a2st — APPROVED** [review]
LGTM

**Pull request merged**